packages/store/src/database.ts

```diff
@@ -68,54 +66,97 @@ export class Database {
+    const eventsByEntityID = new Map<EntityID, Event[]>();
+    for (const event of events) {
+      eventsByEntityID.set(event.entityID, [
+        ...(eventsByEntityID.get(event.entityID) ?? []),
+        event,
+      ]);
+    }
+    const eventsByEntityIDEntries = [...eventsByEntityID.entries()];
+
     const batchSize = 100;
-    for (let i = 0; i < entityIDs.length; i += batchSize) {
-      const currentEntityRecordSnapshots = await this._backend.getEntities(
-        entityIDs,
-      );
-      const newEntityRecordSnapshots = await Promise.all(
-        entityIDs
-          .slice(i, i + batchSize)
-          .map(
-            async (id) =>
-              await this._computeUpdatedEntitySnapshot(
+    for (let i = 0; i < eventsByEntityIDEntries.length; i += batchSize) {
+      const batchEntries = eventsByEntityIDEntries.slice(i, i + batchSize);
+      await this._backend.modifyEntities(
+        batchEntries.map(([id]) => id),
+        async (currentEntityRecordMap) => {
+          const newSnapshotPromises: Promise<
+            DatabaseBackendEntityRecord<Entity>
+          >[] = [];
+          for (const [id, events] of batchEntries) {
+            newSnapshotPromises.push(
+              this._computeUpdatedEntitySnapshot(
                 id,
-                currentEntityRecordSnapshots.get(id) ?? null,
+                currentEntityRecordMap.get(id) ?? null,
+                events,
               ),
-          ),
+            );
+          }
+
+          const newSnapshots = await Promise.all(newSnapshotPromises);
+          return new Map(
+            newSnapshots.map((snapshot) => [snapshot.entity.id, snapshot]),
+          );
+        },
       );
-      await this._backend.putEntities(newEntityRecordSnapshots);
     }
   }
 
   private async _computeUpdatedEntitySnapshot(
     entityID: EntityID,
-    currentRecordSnapshot: DatabaseBackendEntityRecord<Entity> | null,
+    currentRecord: DatabaseBackendEntityRecord<Entity> | null,
+    newEvents: Event[],
   ): Promise<DatabaseBackendEntityRecord<Entity>> {
-    // TODO: attempt to fast-forward by applying events incrementally on the snapshot if possible (i.e. if all the new events' client timestamps are greater than the greatest previous client timestamp)
-    const events = await this._backend.listEvents({
-      predicate: ["entityID", "=", entityID],
-    });
-    const lastEventID = events[events.length - 1];
-    if (events.length > 0) {
+    newEvents.sort((a, b) => a.timestampMillis - b.timestampMillis);
+
+    // Most of the time, new events will be ordered after old events (in terms of local client time), which means we can apply them directly on top of our existing snapshot.
+    if (
+      currentRecord &&
+      newEvents.length > 0 &&
+      newEvents[0].timestampMillis > currentRecord.lastEventTimestampMillis
```

Comment:
Is there the assumption that the client will always send their events sorted by timestamp or are we validating that somewhere else in the stack?

Reply:
Nope: they're sorted on line 115.